File: leetcode/Python/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        c, f = head, head
        while f and f.next:
            c, f = c.next, f.next.next
            if c == f:
                break
        if f is None or f.next is None:
            return None
        c= head
        while c!=f:
            c, f = c.next, f.next
        return c

File: leetcode/C++/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *c=head, *f=head;
        while((f && f->next)){
            c=c->next; f=f->next->next;
            if(c==f) break;
        }
        if(f == nullptr || f->next == nullptr) return nullptr;
        c=head;
        while(c!=f) {c=c->next; f=f->next;}
        return c;
    }
};